ledger/eval/txntracer.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package eval
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+)
+
+// TODO do we want something like this?
+type txnGroupDeltas struct {
+	// Ids contains all the associated IDs for these changes (group and txn IDs)
+	Ids []crypto.Digest
+	ledgercore.StateDelta
+}
+
+// TxnGroupDeltaTracer collects groups of StateDelta objects covering groups of txns
+type TxnGroupDeltaTracer struct {
+	// lookback is the number of rounds stored at any given time
+	lookback uint64
+	// no-op methods we don't care about
+	logic.NullEvalTracer
```

Comment:
Curious what others think about this. There are weird side effects of using inheritance in go like this (i.e. it doesn't actually use polymorphism, calls from the embedded interface don't call the parent overrides). These don't apply to `NullEvalTracer`, but I'm generally hesitant about showing off the pattern.

The alternative is empty function definitions, which doesn't seem _that_ bad.

Reply:
Yeah it does use "inheritance" in ways that don't seem idiomatic. This is how it's being done in other (all other?) tracers that I've seen so far in the codebase. https://github.com/algorand/go-algorand/blob/master/ledger/simulation/tracer.go#L30

Reply:
We considered a few different options before settling on the current `NullEvalTracer` idea. Originally we actually broke up the `EvalTracer` interface into many small ones, since we anticipated most tracers would only want to implement only a few related methods. Having to add a bunch of empty methods for things you don't care about seemed annoying, and it also slowed down development since there were frequent changes to methods.

It's definitely a bit weird though

Reply:
Thanks for the background. You might look into multiple optional interfaces and a cast check. We've been using that pattern for optional functions in Conduit. It's a little more work but a really nice interface to implement.

Here's one example. Plugins can optionally implement the Complete interface for a callback after a round is finished https://github.com/algorand/conduit/blob/585a568f091d333a79b27e872ef781206d90859d/conduit/pipeline/pipeline.go#L73